### PR TITLE
feat(algo-builder/web):early support smart signature in web mode

### PR DESCRIPTION
**Pull request by vuvoth**
## Proposed Changes

+ Early support smart signature transaction in web-mode.ts
+ Simple unit test for web-mode.ts
## Test plan 
On my machine 

Please review it! Thanks

### CODE REVIEW

File: packages/web/src/lib/web-mode.ts [hunk 1]
```diff
@@ -142,23 +143,44 @@ export class WebMode {
...
+		const toBeSignedTxns = base64Txs.map((txn: string, txnId: number) => {
+			return execParams[txnId].sign === types.SignType.LogicSignature
+				? { txn: txn, signers: [] }
+				: { txn: txn };
 		});
+
 		const signedTxn = await this.signTransaction(toBeSignedTxns);
+
+		// sign smart signature transaction
+		for (let i = 0; i < txns.length; ++i) {
```
Comment (rmeena840) [inline packages/web/src/lib/web-mode.ts:164]: Just a note. Try to use below method for iterating to array.

`for (let profile of myArray) {
    console.log(profile.name) // profile is object in myArray
}`

This looks more clean. The advantage of this is that we don't have to calculate the length of array  on every iteration which is done here: `i < txns.length`

**Review by rmeena840 — COMMENTED** [review]
Left comments.

File: packages/web/src/lib/web-mode.ts (hunk 1 above)
Comment (vuvth) [reply packages/web/src/lib/web-mode.ts]: done

File: packages/web/src/lib/web-mode.ts [hunk 2]
```diff
@@ -1,5 +1,6 @@
 import algosdk, { SuggestedParams, Transaction } from "algosdk";
 
+import { types } from "..";
```
[the remark this answers is not in the document]
Comment (vuvth) [reply packages/web/src/lib/web-mode.ts]: fixed

**Review by rmeena840 — COMMENTED** [review]
It's a new featur that you have added. I think it's a good idea to add an unit test of it.

File: packages/web/test/mocks/algo-signer-mock.ts [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+	AlgoSigner,
+	Encoding,
+	JsonPayload,
+	MultisigTransaction,
+	RequestErrors,
+	Transaction,
+	WalletTransaction,
+} from "../../src/algo-signer-types";
+
+const suggestedParamsMock = {
```
Comment (rmeena840) [inline packages/web/test/mocks/algo-signer-mock.ts:15]: Shouldn't we also define the type of this? I think it's suggested params.

File: packages/web/test/src/lib/web-mode.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+	let sender: Account;
+	let receiver: Account;
+
+	this.beforeEach(() => {
+		sender = algosdk.generateAccount();
+		receiver = algosdk.generateAccount();
+		webMode = new WebMode(new AlgoSignerMock(), "Test");
+	});
+
+	it("#executeTx", async function () {
```
Comment (sczembor) [inline packages/web/test/src/lib/web-mode.ts:18]:
```suggestion
	it("Should executeTx without throwing an error", async function () {
```

File: packages/web/test/src/lib/web-mode.ts [hunk 5]
```diff
@@ -0,0 +1,31 @@
+import algosdk, { Account } from "algosdk";
+import assert from "assert";
+
+import { types, WebMode } from "../../../src";
+import { AlgoSignerMock } from "../../mocks/algo-signer-mock";
+
+describe("Webmode test", function () {
```
Comment (sczembor) [inline packages/web/test/src/lib/web-mode.ts:7]: We could be more specific here

File: packages/web/test/mocks/algo-signer-mock.ts [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+	signTxn(transactions: WalletTransaction[], error?: RequestErrors): Promise<JsonPayload> {
+		return new Promise((resolve, reject) => {
+			const result = Array(WallectConnectSession.length);
+			for (let txnId = 0; txnId < transactions.length; ++txnId) {
+				result[txnId] = {
+					txID: "tx-id",
+					blob: "blod",
+				};
+			}
+			return resolve(result as unknown as JsonPayload);
```
Comment (rmeena840) [inline packages/web/test/mocks/algo-signer-mock.ts:106]: What are we trying to do here? First to unknown then JsonPayLoad type. Can't we do directly?

**Review by rmeena840 — CHANGES_REQUESTED** [review]
Reviewed. Left comments.

**Review by sczembor — APPROVED** [review]
Left few comments

File: packages/web/test/mocks/algo-signer-mock.ts (hunk 6 above)
Comment (vuvoth) [reply packages/web/test/mocks/algo-signer-mock.ts:106]: We can't do it directly. TS throw error if we do it.

File: packages/web/test/mocks/algo-signer-mock.ts (hunk 3 above)
Comment (vuvoth) [reply packages/web/test/mocks/algo-signer-mock.ts:15]: Seem likes we can't use suggested params.

**Pull request merged**